File: OpenHome/Net/Bindings/Cpp/ControlPoint/Tests/TestBasicCp.cpp
#include <OpenHome/Types.h>
#include <OpenHome/Net/Cpp/CpOpenhomeOrgTestBasic1.h>
#include <OpenHome/Net/Cpp/CpDevice.h>
#include "TestBasicCp.h"
#include <OpenHome/Private/TestFramework.h>

#include <string>

using namespace OpenHome;
using namespace OpenHome::Net;
using namespace OpenHome::TestFramework;


void TestBasicCp::TestActions(CpDeviceCpp& aDevice)
{
    Print("  Actions\n");
    CpProxyOpenhomeOrgTestBasic1Cpp* proxy = new CpProxyOpenhomeOrgTestBasic1Cpp(aDevice);
    TUint i;

    Print("    Unsigned integer arguments...\n");
    TUint valUint = 15;
    for (i=0; i<kTestIterations; i++) {
        TUint result;
        proxy->SyncIncrement(valUint, result);
        ASSERT(result == valUint+1);
        valUint = result;
    }

    Print("    Integer arguments...\n");
    TInt valInt = 3;
    for (i=0; i<kTestIterations; i++) {
        TInt result;
        proxy->SyncDecrement(valInt, result);
        ASSERT(result == valInt-1);
        valInt = result;
    }

    Print("    Boolean arguments...\n");
    TBool valBool = true;
    for (i=0; i<kTestIterations; i++) {
        TBool result;
        proxy->SyncToggle(valBool, result);
        ASSERT(result == !valBool);
        valBool = result;
    }

    Print("    String arguments...\n");
    std::string valStr("<&'tag\">");
    for (i=0; i<kTestIterations; i++) {
        std::string result;
        proxy->SyncEchoString(valStr, result);
        ASSERT(result == valStr);
    }

    Print("    Binary arguments...\n");
    char bin[256];
    for (i=0; i<256; i++) {
        bin[i] = (char)i;
    }
    std::string valBin(&bin[0], 256);
    for (i=0; i<kTestIterations; i++) {
        std::string result;
        proxy->SyncEchoBinary(valBin, result);
        ASSERT(result == valBin);
        ASSERT(result.length() == 256);
    }

    delete proxy;
}

void TestBasicCp::TestThrowsWhenDisabled(CpDeviceCpp& aDevice)
{
    CpProxyOpenhomeOrgTestBasic1Cpp* proxy = new CpProxyOpenhomeOrgTestBasic1Cpp(aDevice);
    Print("  Actions throw when device disabled\n");
    try {
        TUint valUint;
        proxy->SyncGetUint(valUint);
        ASSERTS();
    }
    catch(ProxyError&) { }
    delete proxy;
}

void TestBasicCp::TestThrowsCustomError(CpDeviceCpp& aDevice)
{
    CpProxyOpenhomeOrgTestBasic1Cpp* proxy = new CpProxyOpenhomeOrgTestBasic1Cpp(aDevice);
    Print("  Custom errors are reported\n");
    try {
        proxy->SyncSetInt(12345);
        ASSERTS();
    }
    catch (ProxyError& pe) {
        ASSERT(pe.Code() == 801);
    }
    delete proxy;
}

static void STDCALL updatesComplete(void* aPtr)
{
    reinterpret_cast<Semaphore*>(aPtr)->Signal();
}

void TestBasicCp::TestSubscriptions(CpDeviceCpp& aDevice)
{
    Semaphore sem("TSEM", 0);
    Print("  Subscriptions\n");
    CpProxyOpenhomeOrgTestBasic1Cpp* proxy = new CpProxyOpenhomeOrgTestBasic1Cpp(aDevice);
    Functor functor = MakeFunctor(&sem, updatesComplete);
    proxy->SetPropertyChanged(functor);
    proxy->Subscribe();
    sem.Wait(); // wait for initial event

    /* For each property,
         call the setter action it
         wait on a property being updated
         check that the property matches the value set
         check that the getter action matches the property
    */

    Print("    Uint...\n");
    proxy->SyncSetUint(1);
    sem.Wait();
    TUint propUint;
    proxy->PropertyVarUint(propUint);
    ASSERT(propUint == 1);
    TUint valUint;
    proxy->SyncGetUint(valUint);
    ASSERT(propUint == valUint);

    Print("    Int...\n");
    proxy->SyncSetInt(-99);
    sem.Wait();
    TInt propInt;
    proxy->PropertyVarInt(propInt);
    ASSERT(propInt == -99);
    TInt valInt;
    proxy->SyncGetInt(valInt);
    ASSERT(propInt == valInt);

    Print("    Bool...\n");
    proxy->SyncSetBool(true);
    sem.Wait();
    TBool propBool;
    proxy->PropertyVarBool(propBool);
    ASSERT(propBool);
    TBool valBool;
    proxy->SyncGetBool(valBool);
    ASSERT(valBool);

    Print("    String...\n");
    std::string str("<&'tag\">");
    proxy->SyncSetString(str);
    sem.Wait();
    std::string propStr;
    proxy->PropertyVarStr(propStr);
    ASSERT(propStr == str);
    // test again to check that PropertyVarStr didn't TransferTo the property
    proxy->PropertyVarStr(propStr);
    ASSERT(propStr == str);
    std::string valStr;
    proxy->SyncGetString(valStr);
    ASSERT(propStr == valStr);

    Print("    Binary...\n");
    char bin[256];
    for (TUint i=0; i<256; i++) {
        bin[i] = (char)i;
    }
    std::string bufBin(&bin[0], 256);
    proxy->SyncSetBinary(bufBin);
    sem.Wait();
    std::string propBin;
    proxy->PropertyVarBin(propBin);
    ASSERT(propBin == bufBin);
    // test again to check that PropertyVarBin didn't TransferTo the property
    proxy->PropertyVarBin(propBin);
    ASSERT(propBin == bufBin);
    std::string valBin;
    proxy->SyncGetBinary(valBin);
    ASSERT(propBin == valBin);

    Print("    Multiple...\n");
    proxy->SyncSetMultiple(15, 658, false);
    sem.Wait();
    proxy->PropertyVarUint(propUint);
    ASSERT(propUint == 15);
    proxy->SyncGetUint(valUint);
    ASSERT(propUint == valUint);
    proxy->PropertyVarInt(propInt);
    ASSERT(propInt == 658);
    proxy->SyncGetInt(valInt);
    ASSERT(propInt == valInt);
    proxy->PropertyVarBool(propBool);
    ASSERT(!propBool);
    proxy->SyncGetBool(valBool);
    ASSERT(!valBool);

    delete proxy; // automatically unsubscribes
}

File: OpenHome/Net/Bindings/Cs/ControlPoint/Tests/TestBasicCp.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Diagnostics;
using System.Linq;
using OpenHome.Net.ControlPoint;
using OpenHome.Net.ControlPoint.Proxies;

namespace OpenHome.Net
{
    class TestBasicCp
    {
        private Semaphore iUpdatesComplete;
        private ICpDevice iDevice;
        private const uint kTestIterations = 10;

        public TestBasicCp(ICpDevice aDevice)
        {
            iDevice = aDevice;
            iUpdatesComplete = new Semaphore(0, Int32.MaxValue);
        }

        public void TestActions()
        {
            Console.Write("  Actions\n");
            CpProxyOpenhomeOrgTestBasic1 proxy = new CpProxyOpenhomeOrgTestBasic1(iDevice);
            uint i;

            Console.Write("    Unsigned integer arguments...\n");
            uint valUint = 15;
            for (i=0; i<kTestIterations; i++) {
                uint result;
                proxy.SyncIncrement(valUint, out result);
                Debug.Assert(result == valUint+1);
                valUint = result;
            }

            Console.Write("    Integer arguments...\n");
            int valInt = 3;
            for (i=0; i<kTestIterations; i++) {
                int result;
                proxy.SyncDecrement(valInt, out result);
                Debug.Assert(result == valInt-1);
                valInt = result;
            }

            Console.Write("    Boolean arguments...\n");
            bool valBool = true;
            for (i=0; i<kTestIterations; i++) {
                bool result;
                proxy.SyncToggle(valBool, out result);
                Debug.Assert(result == !valBool);
                valBool = result;
            }

            Console.Write("    String arguments...\n");
            string valStr = "<&'tag\">";
            for (i=0; i<kTestIterations; i++) {
                string result;
                proxy.SyncEchoString(valStr, out result);
                Debug.Assert(result == valStr);
            }

            Console.Write("    Binary arguments...\n");
            byte[] bin = new byte[128];
            for (i=0; i<127; i++) {
                bin[i] = (byte)(i+1);
            }
            bin[127] = 0;
            for (i=0; i<kTestIterations; i++) {
                byte[] result;
                proxy.SyncEchoBinary(bin, out result);
                Debug.Assert(SequencesEqual(bin, result));
            }

            proxy.Dispose();
        }
        
        // We're not using a version of Mono with .Net 4.0, so we have
        // to implement Zip for ourselves:
        private static IEnumerable<TResult> Zip<TFirst, TSecond, TResult>(
                IEnumerable<TFirst> first,
                IEnumerable<TSecond> second,
                Func<TFirst, TSecond, TResult> resultSelector)
        {
            if (first == null)
            {
                throw new ArgumentNullException("first");
            }
            if (second == null)
            {
                throw new ArgumentNullException("second");
            }
            if (resultSelector == null)
            {
                throw new ArgumentNullException("resultSelector");
            }
            return ZipImpl(first, second, resultSelector);
        }
        private static IEnumerable<TResult> ZipImpl<TFirst, TSecond, TResult>(
                IEnumerable<TFirst> first,
                IEnumerable<TSecond> second,
                Func<TFirst, TSecond, TResult> resultSelector)
        {
            IEnumerator<TFirst> firstEnumerator = first.GetEnumerator();
            IEnumerator<TSecond> secondEnumerator = second.GetEnumerator();
            while (true)
            {
                if (!firstEnumerator.MoveNext())
                {
                    yield break;
                }
                if (!secondEnumerator.MoveNext())
                {
                    yield break;
                }
                yield return resultSelector(firstEnumerator.Current, secondEnumerator.Current);
            }
        }

        private static bool SequencesEqual<T>(IEnumerable<T> xs, IEnumerable<T> ys)
        {
            // We shouldn't have to specify the types to Zip<>, but the
            // Mono 2.6.x compiler has a bug and cannot correctly infer them.
            return Zip<T,T,bool>(xs, ys, EqualityComparer<T>.Default.Equals).All(b=>b);
        }

        public void TestSubscriptions()
        {
            Console.Write("  Subscriptions\n");
            CpProxyOpenhomeOrgTestBasic1 proxy = new CpProxyOpenhomeOrgTestBasic1(iDevice);
            proxy.SetPropertyChanged(UpdatesComplete);
            proxy.Subscribe();
            iUpdatesComplete.WaitOne(); // wait for initial event

            /* For each property,
                 call its setter action
                 wait on a property being updated
                 check that the property matches the value set
                 check that the getter action matches the property
            */

            Console.Write("    Uint...\n");
            proxy.SyncSetUint(1);
            iUpdatesComplete.WaitOne();
            uint propUint = proxy.PropertyVarUint();
            Debug.Assert(propUint == 1);
            uint valUint;
            proxy.SyncGetUint(out valUint);
            Debug.Assert(propUint == valUint);

            Console.Write("    Int...\n");
            proxy.SyncSetInt(-99);
            iUpdatesComplete.WaitOne();
            int propInt = proxy.PropertyVarInt();
            Debug.Assert(propInt == -99);
            int valInt;
            proxy.SyncGetInt(out valInt);
            Debug.Assert(propInt == valInt);

            Console.Write("    Bool...\n");
            proxy.SyncSetBool(true);
            iUpdatesComplete.WaitOne();
            bool propBool = proxy.PropertyVarBool();
            Debug.Assert(propBool);
            bool valBool;
            proxy.SyncGetBool(out valBool);
            Debug.Assert(valBool);

            Console.Write("    String...\n");
            string str = "<&'tag\">";
            proxy.SyncSetString(str);
            iUpdatesComplete.WaitOne();
            string propStr = proxy.PropertyVarStr();
            Debug.Assert(propStr == str);
            // test again to check that PropertyVarStr didn't TransferTo the property
            propStr = proxy.PropertyVarStr();
            Debug.Assert(propStr == str);
            string valStr;
            GC.Collect();
            proxy.SyncGetString(out valStr);
            Debug.Assert(propStr == valStr);

            Console.Write("    Binary...\n");
            byte[] bin = new byte[128];
            for (int i=0; i<127; i++) {
                bin[i] = (byte)(i+1);
            }
            bin[127] = 0;
            proxy.SyncSetBinary(bin);
            iUpdatesComplete.WaitOne();
            byte[] propBin = proxy.PropertyVarBin();
            Debug.Assert(SequencesEqual(propBin, bin));
            // test again to check that PropertyVarBin didn't TransferTo the property
            propBin = proxy.PropertyVarBin();
            Debug.Assert(SequencesEqual(propBin, bin));
            byte[] valBin;
            proxy.SyncGetBinary(out valBin);
            Debug.Assert(SequencesEqual(propBin, valBin));

            Console.Write("    Multiple...\n");
            proxy.SyncSetMultiple(15, 658, false);
            iUpdatesComplete.WaitOne();
            propUint = proxy.PropertyVarUint();
            Debug.Assert(propUint == 15);
            proxy.SyncGetUint(out valUint);
            Debug.Assert(propUint == valUint);
            propInt = proxy.PropertyVarInt();
            Debug.Assert(propInt == 658);
            proxy.SyncGetInt(out valInt);
            Debug.Assert(propInt == valInt);
            propBool = proxy.PropertyVarBool();
            Debug.Assert(!propBool);
            proxy.SyncGetBool(out valBool);
            Debug.Assert(!valBool);

            proxy.Dispose(); // automatically unsubscribes
        }

        private void UpdatesComplete()
        {
            iUpdatesComplete.Release();
        }
    }
}
